File: android-client/app/src/main/java/com/example/awscognito/ui/viewmodel/AuthViewModel.kt
package com.example.awscognito.ui.viewmodel

import android.util.Log
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.amplifyframework.auth.AuthUserAttribute
import com.amplifyframework.auth.AuthUserAttributeKey
import com.amplifyframework.auth.options.AuthSignUpOptions
import com.amplifyframework.core.Amplify
import com.example.awscognito.data.api.ApiClient
import com.example.awscognito.data.model.FeedItem
import com.example.awscognito.data.model.User
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import kotlin.coroutines.resume
import kotlin.coroutines.resumeWithException
import kotlin.coroutines.suspendCoroutine

data class AuthState(
    val isAuthenticated: Boolean = false,
    val isLoading: Boolean = true,
    val userId: String? = null,
    val email: String? = null,
    val error: String? = null,
    val needsConfirmation: Boolean = false,
    val confirmationEmail: String? = null
)

data class DashboardState(
    val user: User? = null,
    val feedItems: List<FeedItem> = emptyList(),
    val isLoading: Boolean = false,
    val error: String? = null
)

class AuthViewModel : ViewModel() {
    private val _authState = MutableStateFlow(AuthState())
    val authState: StateFlow<AuthState> = _authState.asStateFlow()

    private val _dashboardState = MutableStateFlow(DashboardState())
    val dashboardState: StateFlow<DashboardState> = _dashboardState.asStateFlow()

    init {
        checkAuthStatus()
    }

    fun checkAuthStatus() {
        viewModelScope.launch {
            _authState.value = _authState.value.copy(isLoading = true)
            try {
                val session = suspendCoroutine { continuation ->
                    Amplify.Auth.fetchAuthSession(
                        { continuation.resume(it) },
                        { continuation.resumeWithException(it) }
                    )
                }

                if (session.isSignedIn) {
                    val attributes = suspendCoroutine { continuation ->
                        Amplify.Auth.fetchUserAttributes(
                            { continuation.resume(it) },
                            { continuation.resumeWithException(it) }
                        )
                    }

                    val email = attributes.find { it.key == AuthUserAttributeKey.email() }?.value
                    val userId = attributes.find { it.key.keyString == "sub" }?.value

                    _authState.value = AuthState(
                        isAuthenticated = true,
                        isLoading = false,
                        userId = userId,
                        email = email
                    )
                } else {
                    _authState.value = AuthState(isAuthenticated = false, isLoading = false)
                }
            } catch (e: Exception) {
                Log.e(TAG, "Error checking auth status", e)
                _authState.value = AuthState(isAuthenticated = false, isLoading = false)
            }
        }
    }

    fun signUp(email: String, password: String) {
        viewModelScope.launch {
            _authState.value = _authState.value.copy(isLoading = true, error = null)
            try {
                val options = AuthSignUpOptions.builder()
                    .userAttribute(AuthUserAttributeKey.email(), email)
                    .build()

                val result = suspendCoroutine { continuation ->
                    Amplify.Auth.signUp(
                        email,
                        password,
                        options,
                        { continuation.resume(it) },
                        { continuation.resumeWithException(it) }
                    )
                }

                if (result.isSignUpComplete) {
                    _authState.value = _authState.value.copy(
                        isLoading = false,
                        needsConfirmation = false
                    )
                    signIn(email, password)
                } else {
                    _authState.value = _authState.value.copy(
                        isLoading = false,
                        needsConfirmation = true,
                        confirmationEmail = email
                    )
                }
            } catch (e: Exception) {
                Log.e(TAG, "Sign up error", e)
                _authState.value = _authState.value.copy(
                    isLoading = false,
                    error = e.message ?: "Sign up failed"
                )
            }
        }
    }

    fun confirmSignUp(email: String, code: String, password: String) {
        viewModelScope.launch {
            _authState.value = _authState.value.copy(isLoading = true, error = null)
            try {
                suspendCoroutine { continuation ->
                    Amplify.Auth.confirmSignUp(
                        email,
                        code,
                        { continuation.resume(it) },
                        { continuation.resumeWithException(it) }
                    )
                }

                _authState.value = _authState.value.copy(
                    isLoading = false,
                    needsConfirmation = false
                )
                signIn(email, password)
            } catch (e: Exception) {
                Log.e(TAG, "Confirm sign up error", e)
                _authState.value = _authState.value.copy(
                    isLoading = false,
                    error = e.message ?: "Confirmation failed"
                )
            }
        }
    }

    fun signIn(email: String, password: String) {
        viewModelScope.launch {
            _authState.value = _authState.value.copy(isLoading = true, error = null)
            try {
                val result = suspendCoroutine { continuation ->
                    Amplify.Auth.signIn(
                        email,
                        password,
                        { continuation.resume(it) },
                        { continuation.resumeWithException(it) }
                    )
                }

                if (result.isSignedIn) {
                    checkAuthStatus()
                } else {
                    _authState.value = _authState.value.copy(
                        isLoading = false,
                        error = "Sign in incomplete"
                    )
                }
            } catch (e: Exception) {
                Log.e(TAG, "Sign in error", e)
                _authState.value = _authState.value.copy(
                    isLoading = false,
                    error = e.message ?: "Sign in failed"
                )
            }
        }
    }

    fun signOut() {
        viewModelScope.launch {
            _authState.value = _authState.value.copy(isLoading = true)
            try {
                suspendCoroutine { continuation ->
                    Amplify.Auth.signOut { continuation.resume(it) }
                }
                _authState.value = AuthState(isAuthenticated = false, isLoading = false)
                _dashboardState.value = DashboardState()
            } catch (e: Exception) {
                Log.e(TAG, "Sign out error", e)
                _authState.value = _authState.value.copy(isLoading = false)
            }
        }
    }

    fun changePassword(oldPassword: String, newPassword: String, onSuccess: () -> Unit) {
        viewModelScope.launch {
            _authState.value = _authState.value.copy(isLoading = true, error = null)
            try {
                suspendCoroutine { continuation ->
                    Amplify.Auth.updatePassword(
                        oldPassword,
                        newPassword,
                        { continuation.resume(Unit) },
                        { continuation.resumeWithException(it) }
                    )
                }
                _authState.value = _authState.value.copy(isLoading = false)
                onSuccess()
            } catch (e: Exception) {
                Log.e(TAG, "Change password error", e)
                _authState.value = _authState.value.copy(
                    isLoading = false,
                    error = e.message ?: "Password change failed"
                )
            }
        }
    }

    fun loadDashboardData() {
        viewModelScope.launch {
            _dashboardState.value = _dashboardState.value.copy(isLoading = true, error = null)
            try {
                val token = getIdToken()
                if (token == null) {
                    _dashboardState.value = _dashboardState.value.copy(
                        isLoading = false,
                        error = "Not authenticated"
                    )
                    return@launch
                }

                val bearerToken = "Bearer $token"
                val user = ApiClient.apiService.getCurrentUser(bearerToken)
                val feedItems = ApiClient.apiService.getFeed(bearerToken)

                _dashboardState.value = DashboardState(
                    user = user,
                    feedItems = feedItems,
                    isLoading = false
                )
            } catch (e: Exception) {
                Log.e(TAG, "Load dashboard error", e)
                _dashboardState.value = _dashboardState.value.copy(
                    isLoading = false,
                    error = e.message ?: "Failed to load data"
                )
            }
        }
    }

    private suspend fun getIdToken(): String? {
        return try {
            suspendCoroutine { continuation ->
                Amplify.Auth.fetchAuthSession(
                    { session ->
                        val cognitoSession = session as? com.amplifyframework.auth.cognito.AWSCognitoAuthSession
                        val token = cognitoSession?.userPoolTokensResult?.value?.idToken
                        continuation.resume(token)
                    },
                    { continuation.resumeWithException(it) }
                )
            }
        } catch (e: Exception) {
            Log.e(TAG, "Error getting ID token", e)
            null
        }
    }

    fun clearError() {
        _authState.value = _authState.value.copy(error = null)
    }

    companion object {
        private const val TAG = "AuthViewModel"
    }
}

File: ios-client/AWSCognito/AWSCognito/ViewModels/AuthViewModel.swift
//
//  AuthViewModel.swift
//  AWSCognito
//

import Foundation
import Amplify
import AWSCognitoAuthPlugin
import AWSPluginsCore
import Observation

@Observable
@MainActor
final class AuthViewModel {
    // Auth state
    var isAuthenticated = false
    var isLoading = false
    var userId: String?
    var email: String?
    var authError: String?

    // Signup confirmation state
    var needsConfirmation = false
    var confirmationEmail: String?

    // Dashboard state
    var user: User?
    var feedItems: [FeedItem] = []
    var dashboardLoading = false
    var dashboardError: String?

    // Password change state
    var passwordChangeSuccess = false
    var passwordChangeError: String?

    init() {
        Task {
            await checkAuthStatus()
        }
    }

    func checkAuthStatus() async {
        do {
            let session = try await Amplify.Auth.fetchAuthSession()
            isAuthenticated = session.isSignedIn

            if session.isSignedIn {
                await fetchUserAttributes()
            }
        } catch {
            print("Error checking auth status: \(error)")
            isAuthenticated = false
        }
    }

    private func fetchUserAttributes() async {
        do {
            let attributes = try await Amplify.Auth.fetchUserAttributes()
            for attribute in attributes {
                switch attribute.key {
                case .email:
                    email = attribute.value
                case .sub:
                    userId = attribute.value
                default:
                    break
                }
            }
        } catch {
            print("Error fetching user attributes: \(error)")
        }
    }

    func signUp(email: String, password: String) async {
        isLoading = true
        authError = nil

        let userAttributes = [AuthUserAttribute(.email, value: email)]
        let options = AuthSignUpRequest.Options(userAttributes: userAttributes)

        do {
            let result = try await Amplify.Auth.signUp(
                username: email,
                password: password,
                options: options
            )

            switch result.nextStep {
            case .confirmUser(_, _, _):
                needsConfirmation = true
                confirmationEmail = email
            case .done:
                await signIn(email: email, password: password)
            case .completeAutoSignIn:
                isAuthenticated = true
                await fetchUserAttributes()
            }
        } catch let error as AuthError {
            authError = error.errorDescription
        } catch {
            authError = error.localizedDescription
        }

        isLoading = false
    }

    func confirmSignUp(code: String) async {
        guard let email = confirmationEmail else {
            authError = "No email to confirm"
            return
        }

        isLoading = true
        authError = nil

        do {
            let result = try await Amplify.Auth.confirmSignUp(
                for: email,
                confirmationCode: code
            )

            if result.isSignUpComplete {
                needsConfirmation = false
                confirmationEmail = nil
            }
        } catch let error as AuthError {
            authError = error.errorDescription
        } catch {
            authError = error.localizedDescription
        }

        isLoading = false
    }

    func signIn(email: String, password: String) async {
        isLoading = true
        authError = nil

        do {
            let result = try await Amplify.Auth.signIn(
                username: email,
                password: password
            )

            if result.isSignedIn {
                isAuthenticated = true
                await fetchUserAttributes()
            }
        } catch let error as AuthError {
            authError = error.errorDescription
        } catch {
            authError = error.localizedDescription
        }

        isLoading = false
    }

    func signOut() async {
        isLoading = true

        _ = await Amplify.Auth.signOut()
        isAuthenticated = false
        userId = nil
        email = nil
        user = nil
        feedItems = []
        needsConfirmation = false
        confirmationEmail = nil

        isLoading = false
    }

    func changePassword(currentPassword: String, newPassword: String) async {
        isLoading = true
        passwordChangeError = nil
        passwordChangeSuccess = false

        do {
            try await Amplify.Auth.update(oldPassword: currentPassword, to: newPassword)
            passwordChangeSuccess = true
        } catch let error as AuthError {
            passwordChangeError = error.errorDescription
        } catch {
            passwordChangeError = error.localizedDescription
        }

        isLoading = false
    }

    func loadDashboardData() async {
        dashboardLoading = true
        dashboardError = nil

        do {
            guard let token = try await getIdToken() else {
                dashboardError = "Failed to get authentication token"
                dashboardLoading = false
                return
            }

            // Fetch user and feed in parallel
            async let userResult = APIService.shared.fetchCurrentUser(token: token)
            async let feedResult = APIService.shared.fetchFeed(token: token)

            user = try await userResult
            feedItems = try await feedResult
        } catch {
            dashboardError = error.localizedDescription
        }

        dashboardLoading = false
    }

    private func getIdToken() async throws -> String? {
        let session = try await Amplify.Auth.fetchAuthSession()

        if let cognitoTokenProvider = session as? AuthCognitoTokensProvider {
            let tokens = try cognitoTokenProvider.getCognitoTokens().get()
            return tokens.idToken
        }
        return nil
    }
}
